File: n8n-fleet/n8n/agentic/test_realness.py
#!/usr/bin/env python3
"""Tests for the test-realness scorer.

A red->green flip only says something about the source if the test EXERCISED the source. A test that
mocks the class under test, or never touches it, can be driven red and then green entirely by its own
stubbing — the execution evidence is byte-identical to a real proof and means nothing.

The scorer must separate that from legitimate mocking. Mock-heaviness is NOT the signal: the fixtures
below include a real test the pipeline wrote for a resource-leak marker that mocks Connection,
Statement and ResultSet and is completely sound, because the object under test is real and the
verify() is against the real object's behaviour.

    python3 test_realness.py
"""
import json
import os
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, HERE)
import tables  # noqa: E402
tables.SUSPICIONS_TABLE = tables.BUGS_TABLE = "test0000000000"
tables.SCAN_FILES_TABLE = tables.METHOD_RUNS_TABLE = "test0000000000"
import gen_prover  # noqa: E402


def score(src, cls):
    js = """
%s
const r = testRealness(%s, %s);
process.stdout.write(JSON.stringify(r));
""" % (gen_prover.TEST_REALNESS_FN, json.dumps(src), json.dumps(cls))
    with tempfile.NamedTemporaryFile("w", suffix=".js", delete=False) as f:
        f.write(js)
        p = f.name
    try:
        r = subprocess.run(["node", p], capture_output=True, text=True)
        if r.returncode != 0:
            raise AssertionError("scorer threw:\n" + r.stderr[-1500:])
        return json.loads(r.stdout)
    finally:
        os.unlink(p)


FAILED = []


def check(label, cond, detail=""):
    if cond:
        print("  ok   %s" % label)
    else:
        print("  FAIL %s %s" % (label, detail))
        FAILED.append(label)


# --- REAL fixture: written by the pipeline for a HANDLE_LEAK marker. Mock-heavy and SOUND. ---------
SOUND_MOCKY = '''
package org.owasp.webgoat.lessons.sqlinjection.introduction;
import static org.mockito.Mockito.*;
import org.junit.jupiter.api.Test;
public class SqlInjectionLesson10FsmProofTest {
  @Test void statementShouldBeClosedOnAllPaths() throws Exception {
    LessonDataSource ds = mock(LessonDataSource.class);
    Connection conn = mock(Connection.class);
    Statement stmt = mock(Statement.class);
    when(ds.getConnection()).thenReturn(conn);
    when(conn.createStatement(1, 2)).thenReturn(stmt);
    when(stmt.executeQuery(anyString())).thenThrow(new SQLException("boom"));
    SqlInjectionLesson10 lesson = new SqlInjectionLesson10(ds);
    lesson.injectableQueryAvailability("test");
    verify(stmt, atLeastOnce()).close();
  }
}
'''

# --- REAL fixture: static call against a real JDK oracle, no mocks at all. -------------------------
SOUND_STATIC = '''
package org.owasp.webgoat.lessons.challenges.challenge7;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.assertArrayEquals;
public class MD5FsmProofTest {
  @Test void testLargeInputBitCountOverflow() throws Exception {
    byte[] data = new byte[256 * 1024 * 1024];
    byte[] expected = MessageDigest.getInstance("MD5").digest(data);
    byte[] actual = MD5.getHash(data);
    assertArrayEquals(expected, actual, "hash should match");
  }
}
'''

# --- The failure mode being guarded against: the SUBJECT itself is mocked. -------------------------
MOCKS_SUBJECT = '''
import static org.mockito.Mockito.*;
public class ServersFsmProofTest {
  @Test void t() {
    Servers servers = mock(Servers.class);
    when(servers.sort("id")).thenReturn("safe");
    verify(servers, atLeastOnce()).sort("id");
  }
}
'''

# --- The other failure mode: the class under test is never touched at all. -------------------------
NEVER_TOUCHES = '''
import static org.mockito.Mockito.*;
public class ServersFsmProofTest {
  @Test void t() throws Exception {
    Connection conn = mock(Connection.class);
    Statement stmt = mock(Statement.class);
    when(conn.createStatement()).thenReturn(stmt);
    verify(conn, never()).close();
  }
}
'''

# A comment mentioning the class must not count as exercising it.
COMMENT_ONLY = '''
public class ServersFsmProofTest {
  // this test is about Servers.sort() and new Servers(ds)
  @Test void t() { assertTrue(true); }
}
'''


def main():
    print("sound: mock-heavy but drives the real class (real pipeline output)")
    r = score(SOUND_MOCKY, "SqlInjectionLesson10")
    check("sound", r["sound"] is True, str(r["reasons"]))
    check("subject not mocked", r["mocks_subject"] is False)
    check("scores well despite 3 mocks", r["score"] >= 60, str(r["score"]))

    print("\nsound: static call, no mocks (real pipeline output)")
    r2 = score(SOUND_STATIC, "MD5")
    check("sound via static call", r2["sound"] is True, str(r2["reasons"]))
    check("no-stub test scores highest", r2["score"] > r["score"],
          "static=%s mocky=%s" % (r2["score"], r["score"]))

    print("\nunsound: the class under test is itself mocked")
    r3 = score(MOCKS_SUBJECT, "Servers")
    check("flagged as mocking the subject", r3["mocks_subject"] is True)
    check("not sound", r3["sound"] is False)
    check("score is zero", r3["score"] == 0)
    check("reason names the class", any("Servers" in x for x in r3["reasons"]), str(r3["reasons"]))

    print("\nunsound: never touches the class under test")
    r4 = score(NEVER_TOUCHES, "Servers")
    check("not sound", r4["sound"] is False, str(r4["reasons"]))
    check("score is zero", r4["score"] == 0)

    print("\nunsound: only mentioned in a comment")
    r5 = score(COMMENT_ONLY, "Servers")
    check("comment does not count as usage", r5["sound"] is False, str(r5["reasons"]))

    print("\nranking: value assertions outrank interaction-only")
    val = score('public class T { @Test void t(){ Foo f = new Foo(); assertEquals(2, f.calc()); } }', "Foo")
    ver = score('import static org.mockito.Mockito.*;\npublic class T { @Test void t(){ Bar b = mock(Bar.class);'
                ' Foo f = new Foo(b); f.run(); verify(b).close(); } }', "Foo")
    check("both sound", val["sound"] and ver["sound"])
    check("value-asserting test ranks higher", val["score"] > ver["score"],
          "value=%s interaction=%s" % (val["score"], ver["score"]))

    print("\ndegenerate input")
    check("empty source is unsound", score("", "Foo")["sound"] is False)
    check("empty class name is unsound", score(SOUND_STATIC, "")["sound"] is False)

    print()
    if FAILED:
        raise SystemExit("%d check(s) FAILED: %s" % (len(FAILED), ", ".join(FAILED)))
    print("all checks passed")


if __name__ == "__main__":
    main()

File: n8n-fleet/n8n/agentic/src/lib/test-realness.js
'use strict';
/**
 * Is the proof about the REAL code, or about a mock?
 *
 * A red->green flip only says something about the source if the test EXERCISED the source. A test
 * that mocks the class under test, or never touches it, can be driven red and then green entirely by
 * its own stubbing — the execution record is identical to a real proof and means nothing.
 *
 * Mock-heaviness is deliberately NOT the signal. The pipeline's best resource-leak proof mocks
 * Connection, Statement and ResultSet and is completely sound, because the object under test is real
 * and the verify() asserts the real object's behaviour. Mocking a collaborator you cannot stand up is
 * good practice; mocking the subject is the defect.
 */
function testRealness(src, cls) {
  const out = { sound: false, score: 0, reasons: [], mocks_subject: false, touches_real: false };
  const name = String(cls || '').replace(/[^A-Za-z0-9_$]/g, '');
  if (!src || !name) { out.reasons.push('no test source or no class name'); return out; }
  // mask comments and string literals so a class name inside a comment cannot count as a usage
  const s = src
    .replace(/\/\*[\s\S]*?\*\//g, m => m.replace(/[^\n]/g, ' '))
    .replace(/\/\/[^\n]*/g, m => ' '.repeat(m.length))
    .replace(/"(\\.|[^"\\\n])*"/g, '""');
  const has = (re) => re.test(s);
  const count = (re) => (s.match(re) || []).length;
  const N = name.replace(/\$/g, '\\$');

  // THE CARDINAL SIN: the subject itself is a mock, so every observed behaviour is stubbed.
  out.mocks_subject = has(new RegExp('mock\\s*\\(\\s*' + N + '\\s*\\.\\s*class'))
                   || has(new RegExp('@Mock[\\s\\S]{0,80}?\\b' + N + '\\s+\\w'))
                   || has(new RegExp('@(?:Spy|InjectMocks)[\\s\\S]{0,80}?\\b' + N + '\\s+\\w'));

  const constructs = count(new RegExp('new\\s+' + N + '\\s*\\(', 'g'));
  const statics    = count(new RegExp('\\b' + N + '\\s*\\.\\s*[a-z]\\w*\\s*\\(', 'g'));
  out.touches_real = (constructs > 0 || statics > 0) && !out.mocks_subject;

  const asserts = count(/\bassert[A-Z]\w*\s*\(|\bassertThat\s*\(|\bfail\s*\(/g);
  const verifies = count(/\bverify\s*\(/g);
  const stubs = count(/\bwhen\s*\(|\bmock\s*\(|\bgiven\s*\(|@Mock\b/g);

  if (out.mocks_subject) {
    out.reasons.push('the class under test is itself mocked/spied — the test observes stubs, not ' + name);
  }
  if (!constructs && !statics) {
    out.reasons.push('the test never constructs ' + name + ' and never calls a static method on it');
  }
  out.sound = out.touches_real;
  if (!out.sound) return out;

  // Preference ordering among SOUND tests.
  let score = 55;                                   // it does drive the real class
  if (constructs > 0) { score += 20; out.reasons.push('instantiates the real ' + name); }
  else { out.reasons.push('exercises ' + name + ' through static calls'); }
  if (asserts > 0) { score += 15; out.reasons.push(asserts + ' value/state assertion(s)'); }
  if (asserts === 0 && verifies > 0) {
    // Interaction-only is legitimate for "must call close()" style claims, but it is weaker evidence
    // than asserting a value, so it ranks below — it is not penalised into unsoundness.
    score -= 10; out.reasons.push('asserts only on interactions (verify), not on returned values/state');
  }
  if (stubs === 0) { score += 10; out.reasons.push('no stubbing at all — drives the real objects end to end'); }
  else { out.reasons.push(stubs + ' stub/mock setup(s) for collaborators (legitimate when the real ones need a DB/network)'); }
  out.score = Math.max(0, Math.min(100, score));
  return out;
}

/* ---- test exports (stripped when inlined into n8n) ---- */
module.exports = { testRealness };
